tests/lisa/test_energy_model.py

```diff
@@ -302,3 +307,117 @@ def test_get_cpu_capacity(self):
+            """
+            <idle>-0  [000] 0000.0005: cpu_idle:        state=4294967295 cpu_id=0
+            """ # Ramp up everybody's freqs to 2nd OPP
+            """
+            <idle>-0  [000] 0000.0010: cpu_frequency:   state=1500 cpu_id=0
+            <idle>-0  [000] 0000.0010: cpu_frequency:   state=1500 cpu_id=1
+            <idle>-0  [000] 0000.0010: cpu_frequency:   state=4000 cpu_id=2
+            <idle>-0  [000] 0000.0010: cpu_frequency:   state=4000 cpu_id=3
+            """ # Wake up the other CPUs one by one
+            """
+            <idle>-0  [000] 0000.0011: cpu_idle:        state=4294967295 cpu_id=1
+            <idle>-0  [000] 0000.0012: cpu_idle:        state=4294967295 cpu_id=2
+            <idle>-0  [000] 0000.0013: cpu_idle:        state=4294967295 cpu_id=3
+            """ # Put CPU2 into "cluster sleep" (note CPU3 is still awake)
+            """
+            <idle>-0  [000] 0000.0020: cpu_idle:        state=2 cpu_id=2
+            """
+        )
+
+        dir = mkdtemp()
+        path = os.path.join(dir, 'trace.txt')
+        with open(path, 'w') as f:
+            f.write(trace_data)
+        trace = Trace(None, path, ['cpu_idle', 'cpu_frequency'],
+                      normalize_time=False)
+        shutil.rmtree(dir)
+
+        energy_df = em.estimate_from_trace(trace)
+
+        exp_entries = [
+            # Everybody idle
+            (0.0002, {
+                '0': 0.0,
+                '1': 0.0,
+                '2': 0.0,
+                '3': 0.0,
+                '0-1': 5.0,
+                '2-3': 8.0,
+            }),
+            # CPU0 wakes up
+            (0.0005, {
+                '0': 100.0, # CPU 0 now active
+                '1': 0.0,
+                '2': 0.0,
+                '3': 0.0,
+                '0-1': 10.0, # little cluster now active
+                '2-3': 8.0,
+            }),
+            # Ramp freqs up to 2nd OPP
+            (0.0010, {
+                '0': 150.0,
+                '1': 0.0,
+                '2': 0.0,
+                '3': 0.0,
+                '0-1': 15.0,
+                '2-3': 8.0,
+            }),
+            # Wake up CPU1
+            (0.0011, {
+                '0': 150.0,
+                '1': 150.0,
+                '2': 0.0,
+                '3': 0.0,
+                '0-1': 15.0,
+                '2-3': 8.0,
+            }),
+            # Wake up CPU2
+            (0.0012, {
+                '0': 150.0,
+                '1': 150.0,
+                '2': 400.0,
+                '3': 0.0,
+                '0-1': 15.0,
+                '2-3': 40.0, # big cluster now active
+            }),
+            # Wake up CPU3
+            (0.0013, {
+                '0': 150.0,
+                '1': 150.0,
+                '2': 400.0,
+                '3': 400.0,
+                '0-1': 15.0,
+                '2-3': 40.0,
+            }),
+        ]
+
+
+        # We don't know the exact index that will come out of the parsing, so
```

Comment:
I don't quite understand why that's required, can you better explain?
It seems that you want to align the parsed trace events with the above array of expected values, however I cannot see why.

Reply:
Ah yes I'll add some more detail to make this clearer. Basically it's because of `handle_duplicate_index` in TRAPpy.